File: ext/generator-pro.js
import fs from 'fs/promises';
import path from 'path';
import { fileURLToPath } from 'url';
import { titleToCategory } from './titleToCategory.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// ===================================================================
// KONFIGURASI TERPUSAT
// ===================================================================
const CONFIG = {
  rootDir: path.join(__dirname, '..'),
  artikelDir: path.join(__dirname, '..', 'artikel'),
  templateKategori: path.join(__dirname, '..', 'artikel', '-', 'template-kategori.html'),
  masterJson: path.join(__dirname, '..', 'artikel', 'artikel.json'), // Source (Read-Only)
  jsonOut: path.join(__dirname, '..', 'artikel.json'),               // Output (Etalase)
  sitemapTxt: path.join(__dirname, '..', 'sitemap.txt'),            // Cache System

  // Output Sitemaps
  xmlIndexOut: path.join(__dirname, '..', 'sitemap.xml'),
  xmlPostsOut: path.join(__dirname, '..', 'sitemap-1.xml'),
  xmlImagesOut: path.join(__dirname, '..', 'image-sitemap-1.xml'),
  xmlVideosOut: path.join(__dirname, '..', 'video-sitemap-1.xml'),

  xslLink: 'sitemap-style.xsl',
  rssOut: path.join(__dirname, '..', 'rss.xml'),
  baseUrl: 'https://dalam.web.id',
  rssLimit: 30
};

// Daftar Kategori Fisik
const VALID_CATEGORIES = [
  'gaya-hidup', 'jejak-sejarah', 'lainnya', 'olah-media',
'opini-sosial', 'sistem-terbuka', 'warta-tekno'
];

// ===================================================================
// HELPER FUNCTIONS
// ===================================================================
const getMimeType = (url) => {
  if (!url) return 'image/jpeg';
  const ext = url.split('?')[0].split('.').pop().toLowerCase();
  const map = {
    'jpg': 'image/jpeg', 'jpeg': 'image/jpeg', 'png': 'image/png',
    'gif': 'image/gif', 'webp': 'image/webp', 'avif': 'image/avif', 'svg': 'image/svg+xml'
  };
  return map[ext] || 'image/jpeg';
};

const sanitizeTitle = (raw) => raw.replace(/^\p{Emoji_Presentation}\s*/u, '').trimStart();

const slugify = (text) => {
  return text.toString().toLowerCase().trim()
  .replace(/^[^\w\s]*/u, '')
  .replace(/ & /g, '-and-')
  .replace(/[^a-z0-9\s-]/g, '')
  .replace(/\s+/g, '-')
  .replace(/-+/g, '-');
};

const formatISO8601 = (date) => {
  const d = new Date(date);
  return d.toISOString().split('.')[0] + 'Z';
};

const getYoutubeThumb = (url) => {
  const match = url.match(/embed\/([^/?]+)/);
  return match ? `https://img.youtube.com/vi/${match[1]}/hqdefault.jpg` : null;
};

// ===================================================================
// EXTRACTORS
// ===================================================================
const extractTitle = (c) => (c.match(/<title>([\s\S]*?)<\/title>/i)?.[1] || 'Tanpa Judul').trim();
const extractDesc = (c) => (c.match(/<meta\s+name=["']description["'][^>]+content=["']([^"']+)["']/i)?.[1] || '').trim();
const extractPubDate = (c) => c.match(/<meta\s+property=["']article:published_time["'][^>]+content=["']([^"']+)["']/i)?.[1];
const extractImage = (content, filename) => {
  const socialImg = content.match(/<meta\s+[^>]*(?:name|property)=["'](?:og|twitter):image["'][^>]*content=["']([^"']+)["']/i)?.[1];
  return socialImg || `${CONFIG.baseUrl}/img/${filename.replace('.html', '')}.webp`;
};

const extractVideos = (content, title, desc) => {
  const videos = [];
  const iframes = [...content.matchAll(/<iframe[^>]+src=["']([^"']+)["']/gi)];
  iframes.forEach(m => {
    let src = m[1];
    if (src.startsWith('//')) src = 'https:' + src;
      const cleanSrc = src.split('?')[0];
    videos.push({
      loc: cleanSrc,
      title: title,
      description: desc || `Video dari ${title}`,
      thumbnail: getYoutubeThumb(src) || `${CONFIG.baseUrl}/img/default-video.webp`
    });
  });
  return videos;
};

// ===================================================================
// RSS BUILDER
// ===================================================================
const buildRss = (title, items, rssLink, description) => {
  const itemsXml = items.map(it => {
    const mimeType = getMimeType(it.img);
    return `
    <item>
    <title><![CDATA[${it.title}]]></title>
    <link><![CDATA[${it.loc}]]></link>
    <guid><![CDATA[${it.loc}]]></guid>
    <description><![CDATA[${it.desc || sanitizeTitle(it.title)}]]></description>
    <pubDate>${new Date(it.lastmod).toUTCString()}</pubDate>
    <category><![CDATA[${it.category}]]></category>
    <enclosure url="${it.img}" length="0" type="${mimeType}" />
    </item>`;
  }).join('');

  return `<?xml version="1.0" encoding="UTF-8" ?>
  <rss version="2.0" xmlns:atom="http://www.w3.org/2005/Atom">
  <channel>
  <title><![CDATA[${title}]]></title>
  <link><![CDATA[${CONFIG.baseUrl}/]]></link>
  <description><![CDATA[${description}]]></description>
  <language>id-ID</language>
  <atom:link href="${rssLink}" rel="self" type="application/rss+xml" />
  <lastBuildDate>${new Date().toUTCString()}</lastBuildDate>
  ${itemsXml}
  </channel>
  </rss>`;
};

// ===================================================================
// CORE PROCESSOR V8.2 (HYBRID GUARDIAN)
// ===================================================================
const generate = async () => {
  console.log('🚀 Memulai Generator V8.2 (Hybrid Mode: Efisien & Protektif)...');

  try {
    // 1. Baca Cache (Sitemap.txt) & Master JSON
    const sitemapContent = await fs.readFile(CONFIG.sitemapTxt, 'utf8').catch(() => '');
    const processedUrls = new Set(sitemapContent.split('\n').map(l => l.trim()).filter(l => l !== ''));

    const masterRaw = await fs.readFile(CONFIG.masterJson, 'utf8').catch(() => '{}');
    const masterData = JSON.parse(masterRaw);

    // 2. Scan File di Disk
    const filesOnDisk = (await fs.readdir(CONFIG.artikelDir)).filter(f => f.endsWith('.html'));
    const diskSet = new Set(filesOnDisk);

    let finalRootData = {};
    let allItemsFlat = [];
    let processedFilesSet = new Set();
    let validFilesForCleaning = new Set();

    // 3. PROSES DATA DARI MASTER (Read-Only)
    for (const [category, articles] of Object.entries(masterData)) {
      const catSlug = slugify(category);
      for (const art of articles) {
        const fileName = art[1];
        if (diskSet.has(fileName)) {
          const finalUrl = `${CONFIG.baseUrl}/${catSlug}/${fileName.replace('.html', '')}/`;

          if (!finalRootData[category]) finalRootData[category] = [];
          finalRootData[category].push(art);

          processedFilesSet.add(fileName);
          validFilesForCleaning.add(`${catSlug}/${fileName}`);
          allItemsFlat.push({ title: art[0], file: fileName, img: art[2], lastmod: art[3], desc: art[4], category, loc: finalUrl });

          // Cek Cache: Hanya proses HTML jika belum pernah atau ada perubahan
          if (!processedUrls.has(finalUrl)) {
            console.log(`♻️  Mendistribusikan artikel master: ${fileName}`);
            await processAndDistribute(fileName, category, finalUrl);
            processedUrls.add(finalUrl);
          }
        }
      }
    }

    // 4. AUTO-DISCOVERY (File baru yang belum ada di Master)
    for (const file of filesOnDisk) {
      if (!processedFilesSet.has(file)) {
        console.log(`✨ Menemukan file baru: ${file}`);
        const content = await fs.readFile(path.join(CONFIG.artikelDir, file), 'utf8');
        const title = extractTitle(content);
        const category = titleToCategory(title);
        const catSlug = slugify(category);
        const finalUrl = `${CONFIG.baseUrl}/${catSlug}/${file.replace('.html', '')}/`;

        const pubDate = extractPubDate(content) || (await fs.stat(path.join(CONFIG.artikelDir, file))).mtime;
        const newData = [title, file, extractImage(content, file), formatISO8601(pubDate), extractDesc(content)];

        if (!finalRootData[category]) finalRootData[category] = [];
        finalRootData[category].push(newData);

        validFilesForCleaning.add(`${catSlug}/${file}`);
        processedUrls.add(finalUrl);
        await processAndDistribute(file, category, finalUrl, content);

        allItemsFlat.push({ title, file, img: newData[2], lastmod: newData[3], desc: newData[4], category, loc: finalUrl });
      }
    }

    // 5. SMART CLEANING (Proteksi index.html)
    for (const catSlug of VALID_CATEGORIES) {
      const folderPath = path.join(CONFIG.rootDir, catSlug);
      await fs.mkdir(folderPath, { recursive: true });
      const diskFiles = await fs.readdir(folderPath).catch(() => []);

      for (const file of diskFiles) {
        // HANYA hapus file .html yang BUKAN index.html dan tidak ada di data valid
        if (file.endsWith('.html') && file !== 'index.html' && !validFilesForCleaning.has(`${catSlug}/${file}`)) {
          console.log(`🗑️  Menghapus file usang: ${catSlug}/${file}`);
          await fs.unlink(path.join(folderPath, file));
          const urlToRemove = `${CONFIG.baseUrl}/${catSlug}/${file.replace('.html', '')}/`;
          processedUrls.delete(urlToRemove);
        }
      }
    }

    // 6. SORTING & SAVING (Sesuai Logika V7.1)
    allItemsFlat.sort((a, b) => new Date(b.lastmod) - new Date(a.lastmod));
    for (const cat in finalRootData) finalRootData[cat].sort((a, b) => new Date(b[3]) - new Date(a[3]));

    // Update Cache (sitemap.txt) & Root JSON
    await fs.writeFile(CONFIG.sitemapTxt, Array.from(processedUrls).sort().join('\n'));
    await fs.writeFile(CONFIG.jsonOut, JSON.stringify(finalRootData, null, 2));

    // 7. GENERATE XML SITEMAPS & RSS
    let xmlPosts = '', xmlImages = '', xmlVideos = '';
    for (const item of allItemsFlat) {
      xmlPosts += `  <url>\n    <loc>${item.loc}</loc>\n    <lastmod>${item.lastmod}</lastmod>\n  </url>\n`;
      xmlImages += `  <url>\n    <loc>${item.loc}</loc>\n    <lastmod>${item.lastmod}</lastmod>\n    <image:image>\n      <image:loc>${item.img}</image:loc>\n      <image:caption><![CDATA[${item.title}]]></image:caption>\n    </image:image>\n  </url>\n`;

      const content = await fs.readFile(path.join(CONFIG.artikelDir, item.file), 'utf8');
      const vids = extractVideos(content, item.title, item.desc).filter(v => v.loc && !v.loc.includes('${'));
      vids.forEach(v => {
        xmlVideos += `  <url>\n    <loc>${item.loc}</loc>\n    <lastmod>${item.lastmod}</lastmod>\n    <video:video>\n      <video:thumbnail_loc>${v.thumbnail}</video:thumbnail_loc>\n      <video:title><![CDATA[${v.title}]]></video:title>\n      <video:description><![CDATA[${v.description}]]></video:description>\n      <video:player_loc>${v.loc.replace(/&/g, '&amp;')}</video:player_loc>\n    </video:video>\n  </url>\n`;
      });
    }

    const xslHeader = `<?xml version="1.0" encoding="UTF-8"?>\n<?xml-stylesheet type="text/xsl" href="/${CONFIG.xslLink}"?>\n`;
    const latestMod = allItemsFlat[0]?.lastmod || formatISO8601(new Date());

    await fs.writeFile(CONFIG.xmlIndexOut, `${xslHeader}<sitemapindex xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">\n  <sitemap><loc>${CONFIG.baseUrl}/sitemap-1.xml</loc><lastmod>${latestMod}</lastmod></sitemap>\n  <sitemap><loc>${CONFIG.baseUrl}/image-sitemap-1.xml</loc><lastmod>${latestMod}</lastmod></sitemap>\n  <sitemap><loc>${CONFIG.baseUrl}/video-sitemap-1.xml</loc><lastmod>${latestMod}</lastmod></sitemap>\n</sitemapindex>`);
    await fs.writeFile(CONFIG.xmlPostsOut, `${xslHeader}<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">\n${xmlPosts}</urlset>`);
    await fs.writeFile(CONFIG.xmlImagesOut, `${xslHeader}<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9" xmlns:image="http://www.google.com/schemas/sitemap-image/1.1">\n${xmlImages}</urlset>`);
    await fs.writeFile(CONFIG.xmlVideosOut, `${xslHeader}<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9" xmlns:video="http://www.google.com/schemas/sitemap-video/1.1">\n${xmlVideos}</urlset>`);
    await fs.writeFile(CONFIG.rssOut, buildRss('Layar Kosong', allItemsFlat.slice(0, CONFIG.rssLimit), `${CONFIG.baseUrl}/rss.xml`, `Feed artikel terbaru`));

    // 8. LANDING PAGE KATEGORI (Clean URL Logic V6.9)
    const templateHTML = await fs.readFile(CONFIG.templateKategori, 'utf8').catch(() => null);
    if (templateHTML) {
      for (const [cat, articles] of Object.entries(finalRootData)) {
        const slug = slugify(cat);
        const catItems = allItemsFlat.filter(f => f.category === cat);
        const rssUrl = `${CONFIG.baseUrl}/feed-${slug}.xml`;

        // RSS Kategori
        await fs.writeFile(path.join(CONFIG.rootDir, `feed-${slug}.xml`), buildRss(`${cat} - Layar Kosong`, catItems, rssUrl, `Feed kategori ${cat}`));

        // index.html Kategori (Landing Page)
        const icon = cat.match(/(\p{Emoji})/u)?.[0] || '📁';
        const pageContent = templateHTML
        .replace(/%%TITLE%%/g, sanitizeTitle(cat))
        .replace(/%%DESCRIPTION%%/g, `Kumpulan lengkap artikel tentang ${cat}.`)
        .replace(/%%CATEGORY_NAME%%/g, cat)
        .replace(/%%RSS_URL%%/g, rssUrl)
        .replace(/%%CANONICAL_URL%%/g, `${CONFIG.baseUrl}/${slug}/`)
        .replace(/%%ICON%%/g, icon);

        await fs.writeFile(path.join(CONFIG.rootDir, slug, 'index.html'), pageContent);
      }
    }

    console.log(`✅ SELESAI! Root JSON terupdate, Landing Page aman, Master tetap Suci.`);

  } catch (err) {
    console.error('❌ Error Fatal:', err);
  }
};

// Helper untuk Distribusi & Perbaikan HTML
async function processAndDistribute(file, category, finalUrl, preloadedContent = null) {
  const catSlug = slugify(category);
  const destFolder = path.join(CONFIG.rootDir, catSlug);
  await fs.mkdir(destFolder, { recursive: true });

  let content = preloadedContent || await fs.readFile(path.join(CONFIG.artikelDir, file), 'utf8');

  // Fix Canonical & OG URL
  content = content.replace(/<link\s+rel=["']canonical["']\s+href=["'][^"']+["']\s*\/?>/i, `<link rel="canonical" href="${finalUrl}">`)
  .replace(/<meta\s+property=["']og:url["']\s+content=["'][^"']+["']\s*\/?>/i, `<meta property="og:url" content="${finalUrl}">`);

  // Fix Internal Paths (Logika V6.9)
  const fileNameOnly = file.replace('.html', '');
  const oldUrlPattern = new RegExp(`${CONFIG.baseUrl}/artikel/${fileNameOnly}`, 'g');
  content = content.replace(oldUrlPattern, finalUrl);
  content = content.replace(/\/artikel\/-\/([a-z-]+)(\.html)?\/?/g, '/$1/');

  await fs.writeFile(path.join(destFolder, file), content);
}

generate();

File: alat/generator-pro.rs
use anyhow::{Context, Result};
use chrono::{DateTime, Utc};
use rayon::prelude::*;
use regex::{Captures, Regex};
use serde::{Deserialize, Serialize};
use serde_json::json;
use std::collections::{HashMap, HashSet};
use std::fs;
use std::path::{Path, PathBuf};
use std::sync::{Arc, LazyLock, Mutex};

struct CategoryDefinition {
    name: &'static str,
    keywords: &'static [&'static str],
}

// Data kategori persis seperti di JS kamu
static CATEGORIES: &[CategoryDefinition] = &[
    CategoryDefinition {
        name: "Sistem Terbuka",
        keywords: &["apt", "arch", "aur", "bash", "blankon", "bootable", "bsd", "btrfs", "cachyos", "chroot", "compiz", "conky", "cooling", "debian", "desktop", "distro", "dotfiles", "dpkg", "ext4", "fedora", "flatpak", "floorp", "foss", "garuda", "glibc", "gnome", "grub", "hyprland", "kde", "kernel", "komunitas", "kpli", "linux", "lts", "mageia", "mirror", "mx-linux", "nano", "nixos", "open source", "opensuse", "oss", "pacman", "partition", "paru", "perl", "repo", "rescuezilla", "rev", "root", "rsync", "sebarubuntu", "shell", "slackware", "snap", "solaris", "sudo", "systemd", "tar", "ubuntu", "ubuntu party", "usb", "ventoy", "vim", "wayland", "xfce", "yum", "zorin", "zsync"],
    },
CategoryDefinition {
    name: "Olah Media",
    keywords: &["audio", "av1", "batch-rename", "bitrate", "canvas", "codec", "compress", "convert", "deoldify", "dpi", "durasi", "exif", "excel", "ffmpeg", "film", "format", "foto", "framerate", "gabung", "gambar", "ghostscript", "gimp", "grayscale", "h264", "h265", "handbrake", "imagemagick", "iptv", "kompres", "layers", "libreoffice", "metadata", "mewarnai", "mkv", "mp3", "mp4", "multimedia", "ocr", "ogg", "openoffice", "pdftk", "png", "potong", "preset", "rekam", "recursive", "resize", "scan", "split", "spreadsheet", "srt", "subtitle", "svg", "transcribe", "vector", "video", "vlc", "watermark", "webp"],
},
CategoryDefinition {
    name: "Jejak Sejarah",
    keywords: &["adab", "akidah", "al-andalus", "andalusia", "aqidah", "ayub", "baghdad", "bahtera", "baitul-hikmah", "barqa", "bilal", "daulah", "doa", "fatih", "fatwa", "fiqh", "fitnah", "ghibah", "hadis", "haki", "halal", "haram", "hijab", "hijrah", "hijriyah", "hittin", "hukum", "ibnu batutah", "ikhlas", "imam", "iman", "islam", "istighfar", "isra", "janji", "jumat", "khalifah", "khwarizmi", "madinah", "madyan", "masjid", "masyitoh", "maulid", "mesir", "muamalah", "muhammadiyah", "mukjizat", "murad", "musa", "mushaf", "muslim", "nabi", "nuh", "pahlawan", "penaklukan", "perjanjian", "persia", "pertempuran", "piagam", "quran", "qunut", "ramadhan", "risalah", "sabar", "saf", "sahabat", "salam", "salman", "sejarah", "seljuk", "shalat", "shalahuddin", "sirah", "sombong", "sunnah", "surga", "syariat", "tabayun", "tabi'in", "tabut", "tauhid", "tawadhu", "uhud", "umar", "utsman", "utsmaniyah", "yaqub", "yarmuk", "yerusalem", "yusuf", "zaid"],
},
CategoryDefinition {
    name: "Gaya Hidup",
    keywords: &["angkringan", "arabica", "bahagia", "bali", "bandara", "bekapai", "berkendara", "boker", "camilan", "detox", "diet", "e-bpkb", "gaul", "gaya hidup", "gejala", "gerimis", "herbal", "hobi", "hotel", "jagung", "jajanan", "jogja", "kafein", "kesehatan", "kopi", "kerupuk", "kuliner", "kurma", "laundry", "lifestyle", "metode", "minuman", "motor", "ngopi", "niat", "nutrisi", "obat", "ojol", "parkir", "pecel", "pencernaan", "pijat", "psikotes", "resep", "respiro", "robusta", "sakit", "seduh", "sembelit", "service", "sikat", "sparepart", "staycation", "susu", "tidur", "touring", "unboxing", "vixion", "wanita", "wisata"],
},
CategoryDefinition {
    name: "Opini Sosial",
    keywords: &["aci", "adaro", "amanah", "audit", "bisnis", "budaya", "bukalapak", "catatan", "cpns", "cuti", "duit", "ekonomi", "ekspedisi", "etika", "fenomena", "golput", "grobogan", "harian", "hormat", "ibu", "indonesia", "integritas", "iwan fals", "jatos", "jne", "kasih", "kebijakan", "kejujuran", "kepemimpinan", "kerja", "kinerja", "kolom", "kopdar", "korporasi", "kota", "kreativitas", "kritik", "masyarakat", "nostalgia", "opini", "organisasi", "pajak", "pemerintah", "penjilat", "perencanaan", "perjalanan", "perusahaan", "poac", "politik", "ppdb", "produktifitas", "pt", "rencana", "renungan", "sktm", "sosial", "subsidi", "uang", "ujian nasional", "umkm", "viral", "whoosh"],
},
CategoryDefinition {
    name: "Warta Tekno",
    keywords: &["ai", "amd", "android", "api", "automation", "backend", "baterai", "benchmarking", "blogspot", "bootloader", "branch", "browser", "build", "facebook", "canva", "chatgpt", "claude", "cleanup", "cloudflare", "codespaces", "cpu", "curl", "cyber-security", "deep-learning", "dns", "domain", "drive", "encryption", "endpoint", "eula", "firefox", "frontend", "fullstack", "gadget", "gemini", "git", "github", "gitignore", "gorilla glass", "grammarly", "hardware", "hdd", "head", "header", "hosting", "html", "iot", "javascript", "jasper", "jaringan", "json", "js", "kate", "keyring", "laptop", "learning", "lisensi", "llm", "markdown", "meta", "mic", "microsoft exchange", "notion", "npm", "optimasi", "osborne1", "overclock", "pdf", "phishing", "internet", "piracy", "powerbank", "prompt", "pwa", "push", "quickbooks", "refresh", "robots.txt", "samba", "schema", "security", "shutdown", "software", "ssh", "ssh3", "ssd", "ssl", "static site", "sturnus", "tema", "thermal-paste", "thunderbird", "tidio", "tools", "trojan", "virtualbox", "vivaldi", "web", "website", "wifi", "windows", "winget", "wine", "workflow", "yaml", "yml", "foss", "microsoft"],
},
];
// ===================================================================
// KONFIGURASI TERPUSAT
// ===================================================================
const BASE_URL: &str = "https://dalam.web.id";
const ARTIKEL_DIR: &str = "./artikel";
const TEMPLATE_KATEGORI: &str = "./artikel/-/template-kategori.html";
const JSON_OUT: &str = "artikel.json";
const SITEMAP_TXT: &str = "sitemap.txt";
const RSS_LIMIT: usize = 30;

static VALID_CATEGORIES: &[&str] = &[
    "gaya-hidup", "jejak-sejarah", "lainnya", "olah-media",
"opini-sosial", "sistem-terbuka", "warta-tekno"
];

// ===================================================================
// REGEX ENGINE (Compiled Once)
// ===================================================================
static RE_TITLE: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"(?i)<title>([\s\S]*?)</title>").unwrap());
static RE_DESC: LazyLock<Regex> = LazyLock::new(|| Regex::new(r#"(?i)<meta\s+name=["']description["'][^>]+content=["']([^"']+)["']"#).unwrap());
static RE_PUB_DATE: LazyLock<Regex> = LazyLock::new(|| Regex::new(r#"(?i)<meta\s+property=["']article:published_time["'][^>]+content=["']([^"']+)["']"#).unwrap());
static RE_OG_IMAGE: LazyLock<Regex> = LazyLock::new(|| Regex::new(r#"(?i)<meta\s+[^>]*(?:name|property)=["'](?:og|twitter):image["'][^>]*content=["']([^"']+)["']"#).unwrap());
static RE_IFRAME: LazyLock<Regex> = LazyLock::new(|| Regex::new(r#"(?i)<iframe[^>]+src=["']([^"']+)["']"#).unwrap());
static RE_CANONICAL: LazyLock<Regex> = LazyLock::new(|| Regex::new(r#"(?i)<link\s+rel=["']canonical["']\s+href=["'][^"']+["']\s*/?>"#).unwrap());
static RE_OG_URL: LazyLock<Regex> = LazyLock::new(|| Regex::new(r#"(?i)<meta\s+property=["']og:url["']\s+content=["'][^"']+["']\s*/?>"#).unwrap());

// ===================================================================
// DATA STRUCTURES
// ===================================================================
type ArticleTuple = (String, String, String, String, String); // [judul, file, img, tanggal, desc]

#[derive(Serialize, Deserialize, Debug, Clone)]
struct Article {
    title: String,
    file: String,
    img: String,
    lastmod: String,
    desc: String,
    category: String,
    loc: String,
}

// ===================================================================
// HELPERS
// ===================================================================
fn slugify(text: &str) -> String {
    text.to_lowercase().trim()
    .replace(" & ", "-and-")
    .replace(|c: char| !c.is_alphanumeric() && c != ' ', "")
    .split_whitespace()
    .collect::<Vec<_>>()
    .join("-")
}

fn sanitize_title(raw: &str) -> String {
    // Sederhana: hapus emoji di awal jika ada (menggunakan regex jika perlu)
    raw.trim().to_string()
}

// Replikasi logic titleToCategory.js (Sesuaikan dengan file JS kamu)
fn title_to_category(title: &str) -> String {
    let t = title.to_lowercase();

    // Iterasi daftar CATEGORIES yang di atas tadi
    let found = CATEGORIES.iter().find(|cat| {
        // Cek apakah ada keyword yang terkandung dalam judul (t.contains(k))
        cat.keywords.iter().any(|&k| t.contains(k))
    });

    match found {
        Some(cat) => cat.name.to_string(),
        None => "Lainnya".to_string(),
    }
}

// ===================================================================
// CORE PROCESSOR
// ===================================================================
fn main() -> Result<()> {
    println!("🚀 Memulai Rust Generator V8.6 (Tiered Gateway)...");

    // 1. PREPARASI DATA
    let etalase_raw = fs::read_to_string(JSON_OUT).unwrap_or_else(|_| "{}".to_string());
    let etalase_data: HashMap<String, Vec<ArticleTuple>> = serde_json::from_str(&etalase_raw).unwrap_or_default();

    let sitemap_content = fs::read_to_string(SITEMAP_TXT).unwrap_or_default();
    let processed_urls: HashSet<String> = sitemap_content.lines().map(|l| l.trim().to_string()).filter(|l| !l.is_empty()).collect();

    let files_on_disk: Vec<PathBuf> = fs::read_dir(ARTIKEL_DIR)?
    .filter_map(|e| e.ok())
    .map(|e| e.path())
    .filter(|p| p.extension().and_then(|s| s.to_str()) == Some("html"))
    .collect();

    let all_items_flat = Arc::new(Mutex::new(Vec::<Article>::new()));
    let valid_files_cleaning = Arc::new(Mutex::new(HashSet::<String>::new()));
    let final_root_data = Arc::new(Mutex::new(HashMap::<String, Vec<ArticleTuple>>::new()));

    // 2. PEKERJAAN BERTINGKAT (PARALLEL)
    files_on_disk.par_iter().for_each(|path| {
        let file_name = path.file_name().unwrap().to_string_lossy().to_string();
        let mut article_info: Option<(ArticleTuple, String)> = None;

        // --- GATE 1: CEK ETALASE ---
        for (cat, items) in &etalase_data {
            if let Some(found) = items.iter().find(|it| it.1 == file_name) {
                let loc = format!("{}/{}/{}", BASE_URL, slugify(cat), file_name.replace(".html", ""));

                // --- GATE 2: CEK CACHE (sitemap.txt) ---
                if processed_urls.contains(&loc) {
                    article_info = Some((found.clone(), cat.clone()));
                    break;
                }
            }
        }

        // --- GATE 4: AUTO-DISCOVERY ---
        if article_info.is_none() {
            if let Ok(content) = fs::read_to_string(path) {
                let title = RE_TITLE.captures(&content).and_then(|c| c.get(1)).map(|m| m.as_str().to_string()).unwrap_or_else(|| "Tanpa Judul".to_string());
                let category = title_to_category(&title);
                let loc = format!("{}/{}/{}", BASE_URL, slugify(&category), file_name.replace(".html", ""));

                let desc = RE_DESC.captures(&content).and_then(|c| c.get(1)).map(|m| m.as_str().to_string()).unwrap_or_default();
                let date = RE_PUB_DATE.captures(&content).and_then(|c| c.get(1)).map(|m| m.as_str().to_string())
                .unwrap_or_else(|| Utc::now().format("%Y-%m-%dT%H:%M:%SZ").to_string());
                let img = RE_OG_IMAGE.captures(&content).and_then(|c| c.get(1)).map(|m| m.as_str().to_string())
                .unwrap_or_else(|| format!("{}/img/{}.webp", BASE_URL, file_name.replace(".html", "")));

                let new_data = (title.clone(), file_name.clone(), img, date, desc);

                // DISTRIBUSI FILE (Gate 3/4 Logic)
                let _ = process_and_distribute(&file_name, &category, &loc, &content);

                article_info = Some((new_data, category));
            }
        }

        // Simpan Hasil ke Koleksi Global
        if let Some((data, cat)) = article_info {
            let article = Article {
                title: data.0.clone(),
                                      file: data.1.clone(),
                                      img: data.2.clone(),
                                      lastmod: data.3.clone(),
                                      desc: data.4.clone(),
                                      category: cat.clone(),
                                      loc: format!("{}/{}/{}", BASE_URL, slugify(&cat), file_name.replace(".html", "")),
            };

            valid_files_cleaning.lock().unwrap().insert(format!("{}/{}", slugify(&cat), file_name));
            all_items_flat.lock().unwrap().push(article);
            final_root_data.lock().unwrap().entry(cat).or_default().push(data);
        }
    });

    // 3. SMART CLEANING
    for cat_slug in VALID_CATEGORIES {
        let folder_path = Path::new(".").join(cat_slug);
        if let Ok(entries) = fs::read_dir(&folder_path) {
            for entry in entries.filter_map(|e| e.ok()) {
                let fname = entry.file_name().to_string_lossy().to_string();
                if fname.ends_with(".html") && fname != "index.html" {
                    if !valid_files_cleaning.lock().unwrap().contains(&format!("{}/{}", cat_slug, fname)) {
                        println!("🗑️  Menghapus file usang: {}/{}", cat_slug, fname);
                        let _ = fs::remove_file(entry.path());
                    }
                }
            }
        }
    }

    // 4. SORTING & SAVING
    let mut final_items = Arc::try_unwrap(all_items_flat).unwrap().into_inner().unwrap();
    final_items.sort_by(|a, b| b.lastmod.cmp(&a.lastmod));

    let mut root_map = Arc::try_unwrap(final_root_data).unwrap().into_inner().unwrap();
    for items in root_map.values_mut() {
        items.sort_by(|a, b| b.3.cmp(&a.3));
    }

    fs::write(JSON_OUT, serde_json::to_string_pretty(&root_map)?)?;
    fs::write(SITEMAP_TXT, final_items.iter().map(|it| it.loc.clone()).collect::<Vec<_>>().join("\n"))?;

    // 5. LANDING PAGE KATEGORI & RSS
    let template_html = fs::read_to_string(TEMPLATE_KATEGORI).ok();
    if let Some(tpl) = template_html {
        for (cat, articles) in root_map {
            let slug = slugify(&cat);
            let rss_url = format!("{}/feed-{}.xml", BASE_URL, slug);

            // Generate JSON-LD hasPart
            let has_part = articles.iter().map(|a| json!({
                "@type": "WebPage",
                "name": a.0,
                "url": format!("{}/{}/{}", BASE_URL, slug, a.1.replace(".html", "")),
                                                         "datePublished": a.3,
                                                         "description": if a.4.is_empty() { &a.0 } else { &a.4 }
            })).collect::<Vec<_>>();

            let schema_injection = format!(",\n  \"hasPart\": {}", serde_json::to_string_pretty(&has_part)?);

            let page_content = tpl
            .replace("%%TITLE%%", &sanitize_title(&cat))
            .replace("%%CATEGORY_NAME%%", &cat)
            .replace("%%RSS_URL%%", &rss_url)
            .replace("%%CANONICAL_URL%%", &format!("{}/{}", BASE_URL, slug))
            .replace(r#""inLanguage": "id-ID""#, &format!(r#""inLanguage": "id-ID"{}"#, schema_injection));

            let out_dir = Path::new(".").join(&slug);
            fs::create_dir_all(&out_dir)?;
            fs::write(out_dir.join("index.html"), page_content)?;
        }
    }

    println!("✅ SELESAI! Semua Gate terlampaui, performa optimal.");
    Ok(())
}

// ===================================================================
// DISTRIBUTION ENGINE
// ===================================================================
fn process_and_distribute(file: &str, category: &str, final_url: &str, content: &str) -> Result<()> {
    let cat_slug = slugify(category);
    let dest_folder = Path::new(".").join(&cat_slug);
    fs::create_dir_all(&dest_folder)?;

    let mut new_content = content.to_string();

    // Replikasi Regex Replace Canonical & OG
    new_content = RE_CANONICAL.replace(&new_content, |_: &Captures| {
        format!(r#"<link rel="canonical" href="{}">"#, final_url)
    }).to_string();

    new_content = RE_OG_URL.replace(&new_content, |_: &Captures| {
        format!(r#"<meta property="og:url" content="{}">"#, final_url)
    }).to_string();

    // Clean internal links
    let old_path = format!("{}/artikel/{}", BASE_URL, file.replace(".html", ""));
    new_content = new_content.replace(&old_path, final_url);

    // Replace artikel folder links (V8.6 logic)
    let re_folder = Regex::new(r#"/artikel/-/([a-z-]+)(\.html)?/?"#).unwrap();
    new_content = re_folder.replace_all(&new_content, "/$1").to_string();

    fs::write(dest_folder.join(file), new_content)?;
    Ok(())
}
